dpnp/dpnp_algo/dpnp_fill.py

```diff
@@ -0,0 +1,89 @@
+#   this list of conditions and the following disclaimer.
+# - Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+# THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+import dpctl.tensor as dpt
+import dpctl.utils as dpu
+import numpy as np
+from dpctl.tensor._tensor_impl import (
+    _copy_usm_ndarray_into_usm_ndarray,
+    _full_usm_ndarray,
+    _zeros_usm_ndarray,
+)
+
+import dpnp
+
+
+def dpnp_fill(arr, val):
+    arr = dpnp.get_usm_ndarray(arr)
+    exec_q = arr.sycl_queue
+
+    dpnp.check_supported_arrays_type(val, scalar_type=True, all_scalars=True)
+    # if val is an array, process it
+    if dpnp.is_supported_array_type(val):
+        val = dpnp.get_usm_ndarray(val)
+        if val.shape != ():
+            raise ValueError("`val` must be a scalar or 0D-array")
+        if dpu.get_execution_queue((exec_q, val.sycl_queue)) is None:
+            raise dpu.ExecutionPlacementError(
+                "Input arrays have incompatible queues."
+            )
+        a_val = dpt.asarray(
+            val,
+            dtype=arr.dtype,
+            usm_type=arr.usm_type,
+            sycl_queue=exec_q,
+        )
+        a_val = dpt.broadcast_to(a_val, arr.shape)
+        _manager = dpu.SequentialOrderManager[exec_q]
+        dep_evs = _manager.submitted_events
+        h_ev, c_ev = _copy_usm_ndarray_into_usm_ndarray(
+            src=a_val, dst=arr, sycl_queue=exec_q, depends=dep_evs
+        )
+        _manager.add_event_pair(h_ev, c_ev)
+        return
+    elif not dpnp.isscalar(val):
+        raise TypeError(
+            f"Expected `val` to be a 0D-array or Python scalar, got {type(val)}"
+        )
+
+    dt = arr.dtype
+    val_type = type(val)
+    if val_type in [float, complex] and dpnp.issubdtype(dt, dpnp.integer):
+        val = int(val.real)
+    elif val_type is complex and dpnp.issubdtype(dt, dpnp.floating):
+        val = val.real
+    elif val_type is int and dpnp.issubdtype(dt, dpnp.integer):
+        val = np.asarray(val, dtype=dt)[()]
```

Comment:
What is the use case when it is required? Both input array and scalar have integer type, but casting to numpy.ndarray is done.

Reply:
It handles out-of-bounds integers for a data type. This is how dpctl handles it in `full` as well.

```
In [1]: import dpctl.tensor as dpt, dpctl, numpy as np

In [2]: sc = dpt.iinfo(dpt.int32).max + 1

In [3]: # Numpy 2 raises for out-of-bounds integer for type

In [4]: dpt.full((), sc, dtype="i4")
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[4], line 1
----> 1 dpt.full((), sc, dtype="i4")

File ~/repos/dpctl/dpctl/tensor/_ctors.py:1124, in full(shape, fill_value, dtype, order, device, usm_type, sycl_queue)
   1116 dtype = _get_dtype(dtype, sycl_queue, ref_type=type(fill_value))
   1117 res = dpt.usm_ndarray(
   1118     shape,
   1119     dtype=dtype,
   (...)
   1122     buffer_ctor_kwargs={"queue": sycl_queue},
   1123 )
-> 1124 fill_value = _cast_fill_val(fill_value, dtype)
   1126 _manager = dpctl.utils.SequentialOrderManager[sycl_queue]
   1127 # populating new allocation, no dependent events

File ~/repos/dpctl/dpctl/tensor/_ctors.py:1035, in _cast_fill_val(fill_val, dt)
   1033     return fill_val.real
   1034 elif val_type is int and np.issubdtype(dt, np.integer):
-> 1035     return _to_scalar(fill_val, dt)
   1036 else:
   1037     return fill_val

File ~/repos/dpctl/dpctl/tensor/_ctors.py:772, in _to_scalar(obj, sc_ty)
    767 def _to_scalar(obj, sc_ty):
    768     """A way to convert object to NumPy scalar type.
    769     Raises OverflowError if obj can not be represented
    770     using the requested scalar type.
    771     """
--> 772     zd_arr = np.asarray(obj, dtype=sc_ty)
    773     return zd_arr[tuple()]

OverflowError: Python integer 2147483648 out of bounds for int32
```

Reply:
Do you think we can directly call `_cast_fill_val` from dpctl then? (just to avoid duplication)

Reply:
We can, I only avoided here to avoid using a function dpctl considers private, so as long as that's acceptable practice for dpnp I can make the change.

Reply:
I think it would be fine to use private method here, because we did that in other places also.